### PR TITLE
Add sarif reporting for bundle-audit

### PR DESCRIPTION
**Pull request by Kamsiy**
This PR adds SARIF reporting for the bundle-audit scanner by parsing the current bundle-audit output and generates a sarif report

### Sample Bundle Audit Output
```
{
  "info":{
    "vulnerabilities": [{
      "type":"UnpatchedGem", 
      "name":"actionpack", 
      "version":"4.1.15",
      "cve":"CVE-2020-8166",
      "url":"https://groups.google.com/forum/#!topic/rubyonrails-security/NOjKiGeXUgw",
      "advisory_title":"Ability to forge per-form CSRF tokens given a global CSRF token",
      "description":"It is possible to possible to, given a global CSRF token such as the one\npresent in the authenticity_token meta tag, forge a per-form CSRF token for\nany action for that session.\n\nVersions Affected:  rails < 5.2.5, rails < 6.0.4\nNot affected:       Applications without existing HTML injection vulnerabilities.\nFixed Versions:     rails >= 5.2.4.3, rails >= 6.0.3.1\n\nImpact\n------\n\nGiven the ability to extract the global CSRF token, an attacker would be able to\nconstruct a per-form CSRF token for that session.\n\nWorkarounds\n-----------\n\nThis is a low-severity security issue. As such, no workaround is necessarily\nuntil such time as the application can be upgraded.\n",
      "cvss":"",
      "osvdb":"",
      "patched_versions":["~> 5.2.4.3", ">= 6.0.3.1"],
      "unaffected_versions":[]
    }]
  }
}
```

### Generated SARIF output for bundle-audit

```
{
  "version": ...
  "$schema": ...
  "runs": [
    {
      "tool": {
        "driver": {
          "name": "BundleAudit",
          "version": "0.7.0.1",
          "informationUri": "https://github.com/securego/gosec",
          "rules": [
            {
              "id": "CVE-2020-8166",
              "name": "Ability to forge per-form CSRF tokens given a global CSRF token",
              "fullDescription": {
                "text": "Name: actionpack\nVersion: 4.1.15\nDesciption: It is possible to possible to, given a global CSRF token such as the one\npresent in the authenticity_token meta tag, forge a per-form CSRF token for\nany action for that session.\n\nVersions Affected:  rails < 5.2.5, rails < 6.0.4\nNot affected:       Applications without existing HTML injection vulnerabilities.\nFixed Versions:     rails >= 5.2.4.3, rails >= 6.0.3.1\n\nImpact\n------\n\nGiven the ability to extract the global CSRF token, an attacker would be able to\nconstruct a per-form CSRF token for that session.\n\nWorkarounds\n-----------\n\nThis is a low-severity security issue. As such, no workaround is necessarily\nuntil such time as the application can be upgraded.\n"
              },
              "helpUri": "https://groups.google.com/forum/#!topic/rubyonrails-security/NOjKiGeXUgw",
              "help": {
                "text": "More info: https://groups.google.com/forum/#!topic/rubyonrails-security/NOjKiGeXUgw",
                "markdown": "[More info](https://groups.google.com/forum/#!topic/rubyonrails-security/NOjKiGeXUgw)."
              }
            }
          ]
        }
      },
      "results": [
        {
          "ruleId": "CVE-2020-8166",
          "ruleIndex": 0,
          "level": "error",
          "message": {
            "text": "Name: actionpack\nVersion: 4.1.15\nDesciption: It is possible to possible to, given a global CSRF token such as the one\npresent in the authenticity_token meta tag, forge a per-form CSRF token for\nany action for that session.\n\nVersions Affected:  rails < 5.2.5, rails < 6.0.4\nNot affected:       Applications without existing HTML injection vulnerabilities.\nFixed Versions:     rails >= 5.2.4.3, rails >= 6.0.3.1\n\nImpact\n------\n\nGiven the ability to extract the global CSRF token, an attacker would be able to\nconstruct a per-form CSRF token for that session.\n\nWorkarounds\n-----------\n\nThis is a low-severity security issue. As such, no workaround is necessarily\nuntil such time as the application can be upgraded.\n"
          },
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "Gemfile.lock",
                  "uriBaseId": "%SRCROOT%"
                }
              }
            }
          ]
        }
      ],
      "conversion": ....,
      "invocations": ....
    }
  ]
}
```

### CODE REVIEW

File: lib/sarif/bundle_audit_sarif.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
+    def parse_issue(issue)
+      return nil if @urls.include?(issue[:url])
+
+      @urls.add(issue[:url])
+      {
+        id: issue[:cve],
+        name: issue[:advisory_title],
+        level: "MEDIUM",
```
Comment (joshua-ostrom) [inline lib/sarif/bundle_audit_sarif.rb:19]: Be sure to validate that against the schema.  Also as the various levels are used across multiple scanners you might find opportunity to place them in constants.

File: spec/lib/sarif/bundle_audit_spec.rb [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        " will return the underlying \"untrusted\" hash of data that was\nread from the"\
+        " parameters.  Applications that use this return value may be\ninadvertently use"\
+        " untrusted user input.\n\nImpacted code will look something like this:\n\n```\ndef"\
+        " update\n  # Attacker has included the parameter: `{ is_admin: true }`\n  User.update"\
+        "(clean_up_params)\nend\n\ndef clean_up_params\n   params.each { |k, v|  SomeModel.chec"\
+        "k(v) if k == :name }\nend\n```\n\nNote the mistaken use of `each` in the"\
+        " `clean_up_params` method in the above\nexample.\n\nWorkarounds\n-----------\nDo not "\
+        "use the return values of `each`, `each_value`, or `each_pair` in your\napplication.\n"
+
+        expect(bundle_audit_sarif.parse_issue(issue)).to include(
```
Comment (joshua-ostrom) [inline spec/lib/sarif/bundle_audit_spec.rb:29]: Be sure to validate all produced sarif against the json schema -that's our safety net to ensure we are producing valid SARIF.

File: lib/sarif/bundle_audit_sarif.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    BUNDLEAUDIT_URI = 'https://github.com/securego/gosec'.freeze
+
+    def initialize(scan_report)
+      super(scan_report)
+      @logs = @scan_report.to_h[:info][:vulnerabilities]
+      @uri = BUNDLEAUDIT_URI
+      @urls = Set.new
+    end
+
+    def parse_issue(issue)
```
Comment (ghbren) [inline lib/sarif/bundle_audit_sarif.rb:12]: The `cvss` (Common Vulnerability Scoring System) in the bundle audit output is important, although it is sometimes an empty string.
Please add the `cvss` in the sarif.

Comment (ghbren) [reply lib/sarif/bundle_audit_sarif.rb:12]: Also, it is crucial that you add  `patched_versions` and `unaffected_versions` to the sarif.  
If you want, you can just append them to the `fullDescription` for now.

Comment (ghbren) [reply lib/sarif/bundle_audit_sarif.rb:12]: I would also add `type` (like UnpatchedGem) to the start of `fullDescription`.

Comment (ghbren) [reply lib/sarif/bundle_audit_sarif.rb:12]: I'm not sure what `osvdb` values look like, but you can add it to the end of `fullDescription`.

Comment (ghbren) [reply lib/sarif/bundle_audit_sarif.rb:12]: Can you talk to Josh about whether these should be separate attributes?

Comment (joshua-ostrom) [reply lib/sarif/bundle_audit_sarif.rb:11]: Good call outs, I think adding these to a properties makes sense long term.  Short term of including them in the details works for me

**Review by joshua-ostrom — APPROVED** [review]

**Pull request merged**